thread_primitives.h

```diff
@@ -28,65 +28,119 @@
 #ifndef THREAD_PRIMITIVES_H
 #define THREAD_PRIMITIVES_H
 
+#include <condition_variable>
+#include <functional>
+#include <future>
+#include <memory>
+#include <mutex>
+#include <thread>
+
 class ConditionVariablePrivate;
 class MutexPrivate;
 
 class Mutex {
-  public:
-    Mutex();
-	~Mutex();
-    void lock();
-    bool tryLock();
-    void unlock();
-
-    MutexPrivate* impl_;
+public:
+  Mutex() = default;
+  void lock() { mutex_.lock(); }
+  bool tryLock() { return mutex_.try_lock(); }
+  void unlock() { mutex_.unlock(); }
+
+  std::mutex mutex_;
 };
 
 //! Aquire a lock on the mutex passed to the constructor.
 //! The lock is maintained while the ScopedLock object is alive.
 //! The lock is released uppon destruction.
 class ScopedLock {
-  public:
-    ScopedLock(Mutex* mutex) : mutex_(mutex) { mutex->lock(); }
-    ~ScopedLock() { mutex_->unlock(); }
-
-  private:
-    ScopedLock(const ScopedLock&) { }
-    ScopedLock& operator = (const ScopedLock&) { return *this; }
-    Mutex* mutex_;
+public:
+  ScopedLock(Mutex *mutex) : mutex_(mutex) { mutex->lock(); }
+  ~ScopedLock() { mutex_->unlock(); }
+
+private:
+  ScopedLock(const ScopedLock &) {}
+  ScopedLock &operator=(const ScopedLock &) { return *this; }
+  Mutex *mutex_;
 };
 
 class ConditionVariable {
-  public:
-    ConditionVariable();
-    ~ConditionVariable();
-    
-    void signal();
-    void wakeAll();
-    void wait(Mutex* mutex);
-
-  private:
-    ConditionVariablePrivate* impl_;
+public:
+  ConditionVariable() = default;
+
+  void signal() { _condvar.notify_one(); }
+  void wakeAll() { _condvar.notify_all(); }
+  void wait(Mutex *mutex) {
+    std::unique_lock<std::mutex> ulock(mutex->mutex_, std::defer_lock);
+    _condvar.wait(ulock);
+  }
+
+private:
+  std::condition_variable _condvar;
 };
 
 class Thread {
-  public:
-    Thread();
-    ~Thread();
+public:
+  Thread() = default;
+  ~Thread() {
+    if (thread_.joinable()) {
+      thread_.detach();
+    }
+  }
 
-    bool start(int (*func)(void *), void *ptr);
+  bool start(int (*func)(void *), void *ptr) {
 
```

Comment:
why did you put the implementation in the .h file?

Reply:
I though it'd ease reviewing and reading it is easier. It's easy to split later.

Reply:
- [ ] separate impl in .h and .cpp